### PR TITLE
Wishlist

### PR DESCRIPTION
**Pull request by KaustubhM1997**
1) Adding items from product page to wishlist page
2) The item can be removed from wishlist from the product page
3) The item can be removed from wishlist from the wishlist page
4) The number of items in wishlist can be seen in nav

Live preview - https://nozamaa.netlify.app/

### CODE REVIEW

File: src/components/Aside/aside.jsx [hunk 1]
```diff
@@ -3,14 +3,17 @@ import "../../pages/Home/home.css";
...
     <>
       <aside className="filter-container flex-column">
         <div className="filter-header filter-spaces">
           <h3 className="filter-headings">Filters</h3>
-          <a onClick={() => dispatch({type: "CLEAR"})} className="filter-clear" href="#">
+          <a
+            onClick={() => dispatch({ type: "CLEAR" })}
+            className="filter-clear"
+            href="#"
+          >
```
Comment (srishti-maurya) [inline src/components/Aside/aside.jsx]: you may use `Link` from react router here

File: src/components/Wishlist-card/wishlist-card.jsx [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+        });
+      } catch (errors) {
+        console.log(errors);
+      }
+    } else {
+      navigate("/login-page");
+    }
+  };
+
+  console.log(productcard);
```
Comment (srishti-maurya) [inline src/components/Wishlist-card/wishlist-card.jsx]: you may remove `console.log`

File: src/components/Wishlist-card/wishlist-card.jsx [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+            <span className="text-line-through">7,999</span>
+            <span className="discount-percent">40%</span>
+          </p>
+          <button className="btn btn-secondary">Add to Cart</button>
+          <span
+            onClick={() => deleteFromWishlist(productcard._id)}
+            className="card-heart-icon"
+          >
+             <i class="fa-solid fa-xmark"></i>
+             {/* <i class="fa-solid fa-heart"></i> */}
```
Comment (srishti-maurya) [inline src/components/Wishlist-card/wishlist-card.jsx]: you may remove the commented code if not in use

File: src/pages/Product-listing/product-listing.css [hunk 4]
```diff
@@ -64,10 +70,22 @@ input {
...
+  position: relative;
+  display: flex;
+  justify-content: space-between;
 }
 
+.rating-label {
+
+  position: absolute;
+} */
+
```
Comment (srishti-maurya) [inline src/pages/Product-listing/product-listing.css]: I hope this is to be used in the future, if not you may remove it

File: src/contexts/wishlist-context.jsx [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+
+
+    const wishListReducer = (state, action) => {
+
+        console.log(action);
+
+
+        switch (action.type){
+    
+          case "MOVE_TO_WISHLIST":
```
Comment (srishti-maurya) [inline src/contexts/wishlist-context.jsx]: Suggestion: you may use store the action type strings in variables

**Review by srishti-maurya — APPROVED** [review]
I liked the way you have described the code via comments

File: src/components/Aside/aside.jsx (hunk 1 above)
Comment (srishti-maurya) [reply src/components/Aside/aside.jsx]: using anchor tag in React is not a good practise i guess

File: src/components/Aside/aside.jsx [hunk 6]
```diff
@@ -23,96 +26,130 @@ const Aside = () => {
...
-            className="input-checkbox" />
-            <label for="kids"><p className="checkbox-notify">Kids</p></label>
+            <input
+              onChange={(e) =>
+                dispatch({ type: "CATEGORY", payload: e.target.value })
+              }
+              name="category2"
+              checked={state.category.includes("Kids")}
+              type="checkbox"
+              value="Kids"
```
Comment (himadri2110) [inline src/components/Aside/aside.jsx]: The `name` attribute should be more descriptive.

File: src/components/Nav/nav.jsx [hunk 7]
```diff
@@ -1,62 +1,75 @@
...
-          </NavLink>
-        </button>}
-
+        {!Authenticated && (
+          <button className="nav-item button">
+            <NavLink className="login" to="/login-page">
+              Login
+            </NavLink>
+          </button>
+        )}
```
Comment (himadri2110) [inline src/components/Nav/nav.jsx:42]: You can check for authentication and show Login/Logout at the same place. You can use ternary to check for both conditions.

File: src/components/Product-card/product-card.jsx [hunk 8]
```diff
@@ -1,10 +1,96 @@
...
+   const {wishlist} = state;
+
+
+
+  const { title, price, rating, productImg } = productcard; // destrructuring all the keys from productcard
+
+
+  const {auth: {Authenticated, token}} = useAuth();
+
+  const navigate = useNavigate();
```
Comment (himadri2110) [inline src/components/Product-card/product-card.jsx:23]: Suggestion: You can initialize `navigate` in authContext and import it wherever needed.

File: src/components/Product-card/product-card.jsx [hunk 9]
```diff
@@ -1,10 +1,96 @@
...
+    //as soon as the user logs in, the pre-exisiting wishlist items are shown and then anything the user adds from product paage is added below when dispatch fires
+    if (Authenticated) {
+      try {
+        const response = await axios.post(
+          "/api/user/wishlist",
+          {product: productcard}, //product is the predefined key that we pass the value object to
+          {
+            headers: { authorization: token },
+          }
+        );
```
Comment (himadri2110) [inline src/components/Product-card/product-card.jsx]: You can make a service for the API calls.

File: src/components/Wishlist-card/wishlist-card.jsx [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+  } = useAuth();
+  const { dispatch: Wishlistdispatch } = useWishlist(); //renaming dispatch for better readibility
+  const navigate = useNavigate();
+
+  const deleteFromWishlist = async (id) => {
+    if (Authenticated) {
+      try {
+        const response = await axios.delete(`/api/user/wishlist/${id}`, {
+          headers: { authorization: token },
+        });
```
Comment (himadri2110) [inline src/components/Wishlist-card/wishlist-card.jsx:21]: `deleteFromWishlist` function is being called at two places. So you can place it in the context or make a util function for it.

File: src/contexts/wishlist-context.jsx [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+                payload: [...response.data.wishlist],
+              });
+            } catch (errors) {
+              console.log(errors);
+            }
+          } else {
+            dispatch({
+              type: "MOVE_TO_WISHLIST",
+              payload: [],
+            });
```
Comment (himadri2110) [inline src/contexts/wishlist-context.jsx]: Resetting the state on logout is a clever implementation.

File: src/pages/Wishlist/wishlist.js [hunk 12]
```diff
@@ -1,17 +1,28 @@
...
+      <div className="item-heading-wishlist">
+        <h3>My Wishlist</h3>
+      </div>
 
-      <div className="sub-heading title">
-        <h2 className="h2">My Wishlist</h2>
+      <div class="item-list-wishlist">
+        {wishlist.map((item) => {
+          return <WishlistCard productcard={item} />;
+        })}
```
Comment (himadri2110) [inline src/pages/Wishlist/wishlist.js]: You should pass a key prop and give it an id to uniquely identify it.

**Review by himadri2110 — APPROVED** [review]
Good to go.

**Pull request merged**